### DISCUSSION TITLE
automating-microsoft-365-login-mfa-playwright-tests/

### DISCUSSION THREAD

**Discussion by giscus** [estruyf/eliostruyf-comments#625] [Comments]
# automating-microsoft-365-login-mfa-playwright-tests/

Learn how to automate the login process for Microsoft 365 with multi-factor authentication (MFA) using Playwright in this informative blog post.

https://www.eliostruyf.com/automating-microsoft-365-login-mfa-playwright-tests/

**Comment by RuanAlesi** [comment]
Nice! And about Google Login? It looks like even enabling 2FA Google doesn't allow it. It's a shame considering more and more apps are using Google Login to authenticate the users.

**Reply by estruyf** [reply]
Hi @RuanAlesi, I have yet to try out MFA-enabled Google solutions. What I noticed from a quick test is that it sometimes tries to call or send a text to your mobile after inserting the OTP code. Which makes it tricky to automate.

**Comment by Krugerro** [comment]
Very nice article. Many thanks for taking this subject. There are lots of sources where E2E with MFA questions hangs without answer.👍

I followed your guide and stacked on code verification. Code is generated but it is wrong and I got "you didn't enter expected verification code. Please try again". 

Do you happen to know what I made wrong?

**Reply by estruyf** [reply]
Thanks for the feedback @Krugerro. It might be the secret that was not correct, or you generated it just when the key should have been rotated.

You can always verify the OTP code if you also register it in your authenticator app on your phone. It should give you the same code.

**Reply by Krugerro** [reply]
My mistake. Secret was right but I did not reset my terminal after setting environmental variable.

Many thanks for fast reply.

**Comment by cjwest13** [comment]
This is such a helpful guide. I appreciate you taking the time to do this! I ran into an issue when adding the TOTP. Seems like when I go through the flow of adding the authentication app, the secret key and account name is not present. Instead it’s a code and url, which seems not to represent a secret key. Do you have any insight on what to do in this situation?

**Reply by estruyf** [reply]
@cjwest13, if you get that, you will get the MS Auth details when you go back to the dialog and click on the `I want to use a different authenticator app` link.

![image](https://github.com/user-attachments/assets/0b2c8c89-7899-4402-b85c-9c7b0b79e03f)

**Reply by cjwest13** [reply]
Unfortunately, I don’t have that option. Our company is enforcing the use of Microsoft authenticator only.

**Reply by cjwest13** [reply]
So the line you highlight is not presented in my view.

**Reply by estruyf** [reply]
In that case, the only option you have is to generate an authenticated session and use that until it expires. You can do that via the Playwright codegen tool: https://playwright.dev/docs/codegen#preserve-authenticated-state

**Reply by estruyf** [reply]
@cjwest13 I wrote down the approach: https://www.eliostruyf.com/e2e-testing-mfa-environment-playwright-auth-session/

**Comment by Adilaver** [comment]
Hey!
I am trying with this implementation in our project and it is working, but the problem is the OTP it is generating is different than authenticator.
Getting an error that not a expected OTP.
The nody everything i passed is same with secret key.

**Reply by estruyf** [reply]
You can validate it by entering the secret key in your authenticator app. The OTP code should be the same as your authenticator app and on your computer. OTP only uses the time + secret to generate the code, so it is not device-specific.

**Reply by Adilaver** [reply]
No its not same.
Even i tried generating it through command but its different than app.

**Reply by estruyf** [reply]
That's odd.

I just tried it on my end, and both generate the same code.

**MS Authenticator**

![image](https://github.com/user-attachments/assets/5234a54e-2a9e-4c1c-b8d3-062941b84b88)

**Script**

<img width="468" alt="image" src="https://github.com/user-attachments/assets/7083b7cb-ac29-4f5c-89b7-ec1fe6b5c0a4" />

**Reply by estruyf** [reply]
If you don't find a solution, another way could be to use the auth session: https://www.eliostruyf.com/e2e-testing-mfa-environment-playwright-auth-session/

**Reply by Adilaver** [reply]
Is it possible to connect over LinkedIn? I sent request.

**Reply by estruyf** [reply]
Accepted.

**Reply by Adilaver** [reply]
Hey It worked today, strange!

**Comment by MartinJohns** [comment]
Your example does not work for me. Specifically this case:

```typescript
  // 6. Check if the account has the Microsoft Authenticator app configured
  const otherWayLink = page.locator("a#signInAnotherWay");
  await otherWayLink.waitFor({ timeout: 2000 });
  if (await otherWayLink.isVisible()) { /* ... */ }
```

The `waitFor()` call throws a `TimeoutError` in case of a timeout, so it's not actually possible to check if the element is visible.

**Reply by estruyf** [reply]
> The waitFor() call throws a TimeoutError in case of a timeout, so it's not actually possible to check if the element is visible.

Could you show me the error?

How is your auth configured? With MFA or without?

**Comment by kamalrao091** [comment]
Very nice article. I have followed every step and every thing very smooth and code getting generated but after entering code I am facing couple of issues 
1. I am getting push notification to Authenticator app 
2. I am getting 'You didn't enter the expected verification code. Please try again.' in 
Screenshot 2025-11-02 at 9.01.44 AM
IMG_1018.jpeg

**Reply by kamalrao091** [reply]
Looks like the code that my script generating is different when I compared with what I am getting in authenticator app in mobile. When I manually used entered the code that my authenticator app generated then it is working fine. Any idea why ?

**Reply by estruyf** [reply]
The only way this can happen is when the secret is not the same as in your authenticator app. That might be something you need to verify. In the end, both should give the same code, as it uses the secret and current time to create the 6-digit code.

**Reply by kamalrao091** [reply]
Thanks it worked. Actually I am not reading my secret in the script properly. Is there any doc or info available that we can automate this login using API instead of UI ?
